DuckDuckGo/MainViewController.swift

```diff
@@ -1774,4 +1788,49 @@ extension MainViewController: VoiceSearchViewControllerDelegate {
     }
 }
 
+
+// MARK: - History UIMenu Methods
+
+extension MainViewController {
+
+    private func refreshBackForwardMenuItems() {
+        guard let currentTab = currentTab, #available(iOS 14.0, *) else {
+            return
+        }
+        omniBar.backButton.menu = historyMenu(with: currentTab.webView.backForwardList.backList.reversed())
+        backButton.menu = historyMenu(with: currentTab.webView.backForwardList.backList.reversed())
+        
+        omniBar.forwardButton.menu = historyMenu(with: currentTab.webView.backForwardList.forwardList)
+        forwardButton.menu = historyMenu(with: currentTab.webView.backForwardList.forwardList)
+    }
+
+    @available(iOS 14.0, *)
+    private func historyMenu(with backForwardList: [WKBackForwardListItem]) -> UIMenu {
+        let historyItemList = backForwardList.map { BackForwardMenuHistoryItem(backForwardItem: $0) }
+        let actions = historyMenuButton(with: historyItemList)
+        return UIMenu(title: "", children: actions)
+    }
+    
+    @available(iOS 14.0, *)
+    private func historyMenuButton(with menuHistoryItemList: [BackForwardMenuHistoryItem]) -> [UIAction] {
+        let menuItems: [UIAction] = menuHistoryItemList.compactMap { historyItem in
+            
+            if #available(iOS 15.0, *) {
+                return UIAction(title: historyItem.title,
+                                subtitle: historyItem.sanitizedURLForDisplay,
+                                discoverabilityTitle: historyItem.sanitizedURLForDisplay) { [weak self] _ in
+                    self?.loadBackForwardItem(historyItem.backForwardItem)
+                }
+            } else {
+                return  UIAction(title: historyItem.title,
+                                 discoverabilityTitle: historyItem.sanitizedURLForDisplay) { [weak self] _ in
+                    self?.loadBackForwardItem(historyItem.backForwardItem)
+                }
+            }
+        }
+        
+        let maxNumberOfItems = 6
```

Comment:
Since we're capping the maximum number of items, what do you think about doing this truncation at the start of the function before we map items to `UIAction`? We can avoid unnecessary allocations that way 🙂

Reply:
This limitation should not be there. Safari doesn't have it, it just adds a scroll to the Menu. Good catch, thanks